core/src/main/scala/org/apache/spark/deploy/history/HistoryServer.scala

```diff
@@ -150,14 +150,18 @@ class HistoryServer(
       ui: SparkUI,
       completed: Boolean) {
     assert(serverInfo.isDefined, "HistoryServer must be bound before attaching SparkUIs")
-    ui.getHandlers.foreach(attachHandler)
+    handlers.synchronized {
+      ui.getHandlers.foreach(attachHandler)
+    }
     addFilters(ui.getHandlers, conf)
```

Comment:
shall we also keep `handlers` synchronized for this line?

Reply:
Yes. Thanks for catching this. Will update.